src/test/java/com/github/packageurl/PackageURLBuilderTest.java

```
@@ -95,26 +95,22 @@ public void testPackageURLBuilder() throws MalformedPackageURLException {
 
     @Test
     public void testPackageURLBuilderException1() throws MalformedPackageURLException {
-        exception.expect(MalformedPackageURLException.class);
-        exception.expectMessage("contains a qualifier key with an empty or null");
         PackageURL purl = PackageURLBuilder.aPackageURL()
                 .withType("type")
                 .withName("name")
                 .withQualifier("key","")
                 .build();
-        Assert.fail("Build should fail due to invalid qualifier (empty value)");
+        assertNull(purl.getQualifiers());
```

This test case should probably remain unchanged: `build()` should fail instead of silently ignoring the invalid key.

---

No, because empty keys are legal (as input), so the point is that it _should not_ fail there, just drop the entry.

The builder does no input validation, so `withQualifier(key, "")` is the same as `withoutQualifier(key)` once the purl gets constructed.

Here is what the spec says:

> `value` cannot be an empty string: a `key=value` pair with an empty `value` is the same as no key/value at all for this key.

The spec is unclear and seems to be contradicting itself: it uses the word _cannot_,  but this must be referring to _output_ and not _input_, as we see later on:

> Discard any key/value pairs where the value is empty

Therefore, I interpret this as a key with no value being allowed in input and then (silently) dropped from the output/canonical form.

So this pull request fixes two things:

1. Keys are allowed as input in any case and are silently converted to lowercase for output.
2. Empty keys are allowed as input and are silently dropped from output.